File: Projetos/aeds2/lab/aquecimento-prova-2/C - Concurso de Levantamento de Peso/weight-competition.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_NAME_LENGTH 50
#define MAX_LINE_LENGTH 100
#define MAX_ARRAY 100
#define MAX_WEIGHT 500

typedef struct {
  char name[MAX_NAME_LENGTH];
  unsigned short maxLifted;
} Atlete;

Atlete* parseAtlete(char* input) {
  Atlete* atlete = malloc(sizeof(Atlete));
  if (atlete == NULL) return NULL;

  int i, j;
  for (i = 0; input[i] != ' '; i++) atlete->name[i] = input[i];

  i++;
  char maxLifted[4];
  for (j = 0; input[i] != '\n'; i++, j++) maxLifted[j] = input[i];
  atlete->maxLifted = atoi(maxLifted);

  return atlete;
}

void quicksort(Atlete* atletes[], int left, int right);

void sort(Atlete* atletes[], int competitors) {
  quicksort(atletes, 0, competitors - 1);
}

int main() {
  Atlete* atletes[MAX_ARRAY];
  int competitors = 0;

  char line[MAX_LINE_LENGTH];
  while (fgets(line, MAX_LINE_LENGTH, stdin)) {
    atletes[competitors++] = parseAtlete(line);
  }

  sort(atletes, competitors);

  for (int i = 0; i < competitors; i++) {
    printf("%s %d\n", atletes[i]->name, atletes[i]->maxLifted);
    free(atletes[i]);
  }

  return 0;
}

void swap(Atlete* array[], int firstIndex, int secondIndex) {
  Atlete* temp = array[firstIndex];
  array[firstIndex] = array[secondIndex];
  array[secondIndex] = temp;
}

void quicksort(Atlete* atletes[], int left, int right) {
  int i = left, j = right;
  char* pivotName = atletes[(i + j) / 2]->name;
  int pivotMaxLifted = atletes[(i + j) / 2]->maxLifted;

  while (i <= j) {
    while (atletes[i]->maxLifted > pivotMaxLifted
      || (atletes[i]->maxLifted == pivotMaxLifted
        && strcmp(atletes[i]->name, pivotName) < 0)
      ) i++;

    while (atletes[j]->maxLifted < pivotMaxLifted
      || (atletes[j]->maxLifted == pivotMaxLifted
        && strcmp(atletes[j]->name, pivotName) > 0)
      ) j--;

    if (i <= j) swap(atletes, i++, j--);
  }

  if (left < j) quicksort(atletes, left, j);
  if (i < right) quicksort(atletes, i, right);
};

File: Matérias/aeds2/lab/aquecimento-prova-2/JAVA - Concurso de Levantamento de Peso/WeightCompetition.java

import java.util.Scanner;

class Atlete {

    final private String name;
    final private int maxLifted;

    public Atlete(String name, int maxLifted) {
        this.name = name;
        this.maxLifted = maxLifted;
    }

    public String getName() {
        return name;
    }

    public int getMaxLifted() {
        return maxLifted;
    }

    @Override
    public String toString() {
        return name + " " + maxLifted;
    }
}

public class WeightCompetition {

    static Atlete[] atletes;

    static Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) {
        int competitors = Integer.parseInt(scanner.nextLine());
        atletes = new Atlete[competitors];

        for (int competitor = 0; competitor < competitors; competitor++) {
            String[] line = scanner.nextLine().split(" ");
            String name = line[0];
            int maxLifted = Integer.parseInt(line[1]);

            atletes[competitor] = new Atlete(name, maxLifted);
        }

        sort();
        for (Atlete atlete : atletes) {
            System.out.println(atlete.toString());
        }
    }

    static void sort() {
        quicksort(0, atletes.length - 1);
    }

    static void quicksort(int left, int right) {
        int i = left, j = right;
        String pivotName = atletes[(i + j) / 2].getName();
        int pivotMaxLifted = atletes[(i + j) / 2].getMaxLifted();

        while (i <= j) {
            while (atletes[i].getMaxLifted() > pivotMaxLifted
                    || (atletes[i].getMaxLifted() == pivotMaxLifted
                    && atletes[i].getName().compareToIgnoreCase(pivotName) < 0)) {
                i++;
            }

            while (atletes[j].getMaxLifted() < pivotMaxLifted
                    || (atletes[j].getMaxLifted() == pivotMaxLifted
                    && atletes[j].getName().compareToIgnoreCase(pivotName) > 0)) {
                j--;
            }

            if (i <= j) {
                swap(i++, j--);
            }
        }

        if (j > left) {
            quicksort(left, j);
        }

        if (i < right) {
            quicksort(i, right);
        }
    }

    static void swap(int firstIndex, int secondIndex) {
        Atlete temp = atletes[firstIndex];
        atletes[firstIndex] = atletes[secondIndex];
        atletes[secondIndex] = temp;
    }
}
